database/raft/raft.go

```diff
@@ -0,0 +1,920 @@
+// Package raft provides a simple key-value store coordinated
+// across a raft cluster.
+package raft
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"hash/crc32"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime/debug"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/coreos/etcd/snap/snappb"
+	"github.com/coreos/etcd/wal"
+	"github.com/coreos/etcd/wal/walpb"
+
+	"chain/database/raft/state"
+	"chain/errors"
+	"chain/log"
+)
+
+var ErrUnsatisfied = errors.New("precondition not satisfied")
+
+// TODO(kr): do we need a "client" mode?
+// So we can have many coreds without all of them
+// having to be active raft nodes.
+// (Raft isn't really meant for more than a handful
+// of consensus participants.)
+
+const (
+	tickDur           = 100 * time.Millisecond
+	electionTick      = 10
+	heartbeatTick     = 1
+	maxRaftReqSize    = 10e6 // 10MB
+	snapCount         = 10000
+	dummyWriteTimeout = 50 * time.Millisecond
+
+	nSnapCatchupEntries uint64 = 10000
+
+	contentType = "application/octet-stream"
+)
+
+var crcTable = crc32.MakeTable(crc32.Castagnoli)
+
+// Service holds the key-value data and performs raft coordination.
+type Service struct {
+	// config
+	dir     string
+	id      uint64
+	mux     *http.ServeMux
+	rctxReq chan rctxReq
+	wctxReq chan wctxReq
+	donec   chan struct{}
+
+	errMu sync.Mutex
+	err   error
+
+	confChangeID uint64 // atomic access only
+
+	// The storage object is purely for internal use
+	// by the raft.Node to maintain consistent persistent state.
+	// All client code accesses the cluster state via our Service
+	// object, which keeps a local, in-memory copy of the
+	// complete current state.
+	raftNode    raft.Node
+	raftStorage *raft.MemoryStorage
+
+	// The actual replicated data set.
+	stateMu   sync.Mutex
+	stateCond sync.Cond
+	state     *state.State
+	done      bool
+
+	// Current log position.
+	// access only from runUpdates goroutine
+	snapIndex uint64
+	confState raftpb.ConfState
```

Comment:
In general, it would be good to run it once under -race and see if it warns about anything.